README.md
# SP0256
Allophone speech synthesis in Pure Perl. Multiple methods of getting Perl applications to speak are available.  As far as I can tell all use external libraries to do the heavy lifting. I want to try and implement one with near-zero external dependencies, mainly for simple recognisable speech audio in games etc. Thought I'd start with the classic SP0256 chip as a starting point in this exploration. 

This modulino includes a transformation engine that translates text to allophones.  

[SP0256.webm](https://user-images.githubusercontent.com/34284663/199564715-6f7166c4-04fb-4ff9-876b-60613887959c.webm)

The allophones are rather distorted, and the source I used (from a Github repo of ExtremeEectronis) appears to be also derived from an innaccurate source acording to the author.  It would be an idea to evaluate how different from natural sounds these allophones are, certainly there are allophones in thgis set that, at least to my ear do not sound quite the way one might expect.  Hence a console based audio analyser...

[AudioAnalyser.webm](https://user-images.githubusercontent.com/34284663/200185500-67eed845-cce0-4195-b6c1-81937a972edf.webm)

Note: this Analyser needs Term::Graille v0.15 at least

### DISCLAIMER

I am not a expert in phonetics, linguistics or audio data analysis.  I also happen to be more tone deaf than an earwig with no ears.  So apologies for any offense I cause with my lack of knowledge and insight in this field.

### Changes 

* version 0.03 has an experimental Windows version requiring Win32::Sound 

### Reading List
[Sample rate conversion](https://www.psaudio.com/copper/article/sample-rate-conversion/)

[A Source of C  Allophone data](https://github.com/ExtremeElectronics/SP0256-AL2-Pico-Emulation-Detail)  This is derived form a source which is known to be innaccurate;  a more accurate allophone set would be nice

[An IP Telephone](https://www.foo.be/docs/tpj/issues/vol5_3/tpj0503-0002.html)

[PLOYTEC](https://www.ploytec.com/pl2/pl02_56_release_notes.pdf) has a useful dictionary and gives a good insight on how to maximise the yield with the chip.

[p5-NRL-TextToPhoneme](https://github.com/greg-kennedy/p5-NRL-TextToPhoneme) by [Greg Kennedy](https://greg-kennedy.com/) 

[Original ROMs](https://k1.spdns.de/Vintage/Sinclair/82/Peripherals/Currah%20uSpeech/Technical%20Information/)

[Allophonic Rules](https://youtu.be/vaRTzIoEp9k)

[Software waveforms](https://swphonetics.com/praat/tutorials/understanding-waveforms/speech-waveforms)

[Waveforms of Speech](https://www.mq.edu.au/about/about-the-university/our-faculties/medicine-and-health-sciences/departments-and-centres/department-of-linguistics/our-research/phonetics-and-phonology/speech/acoustics/speech-waveforms/the-waveforms-of-speech) by [Robert Mannell](https://www.researchgate.net/profile/Robert-Mannell-2) at MacQuarie University



CHANGES
v 0.01  : 2022-11-02  Initial experimental release, with blog and reddit announcement
v 0.02  : 2022-11-03  Increase dictionary size with numbers, weekdays and months
                      Allow integers upto 10**10-1 to be spoken
                      start populating some rules
v 0.03  : 2022-11-05  Attempt at Windows compatibility (Depends on Win32::Sound


SP0256/lib/Speech/SP0256/Rules.pm
package Speech::SP0256::Rules;

sub setRules{
	my ($self)=@_;
	$self->{rules}=[
	   ["dge\$", "JH"    ],
	   ["le\$" ,  "EL"   ],
	   ["^i\$" , "AY"    ],
	   ["^un"  , "AA NN" ],
       ["ght"  , "TT1"   ],
	   ["age"  ,  "EY JH"],
	   ["ck"   ,  "KK"   ],
	   ["ch"   , "CH"    ],
	   ["ng"   ,  "NG"   ],
	   ["ea"   ,  "Ih IH"],
	   ["^d"   ,  "DD2"  ],
	   
	   # for those that don't fit, the remaining letters are
	   # directly translated
	   [ "a","EH"],
		["b","BB1"],
		["c","KK1"],
		["d","DD1"],
		["e","EH"],
		["f","FF"],
		["g","GG1"],
		["h","HH1"],
		["i","IH"],
		["j","JH"],
		["k"," KK1 "],
		["l","LL"],
		["m","MM"],
		["n","NN1"],
		["o","OW"],
		["p","PP"],
		["q","KK2"],
		["r","RR1"],
		["s","SS"],
		["t","TT1"],
		["u","UW1"],
		["v","VV"],
		["w","WW"],
		["x","KK2 SS"],
		["y","IH"],
		["z","ZZ"],
	 ]
	   
	   
}

sub addRule{
	my ($self,$match,$result)=@_;
	$self->{rules}=[sort ruleSort([$match,$result],@{$self->{rules}})];
}

sub ruleSort{
	if (length $a->[0] > length $b->[0]) {return -1}
	elsif (length $a->[0] == length $b->[0]){return 0}
	else  {return 1}	
}


1;


SP0256/lib/Speech/SP0256/Dictionary.pm
package Speech::SP0256::Dictionary;

sub initDic{
	my $self=shift;
	$self->{dic}={
		" "=>"PA3",
		","=>"PA4",
		"."=>"PA5",
		0=>"ZZ YR OW",
		1=>"WW SX AX NN1",
		2=>"TT2 UW2",
		3=>"TH RR1 IY",
		4=>"FF OR",
		5=>"FF AY VV",
		6=>"SS SS IH IH PA3 KK2 SS",
		7=>"SS SS EH EH VV IH NN1",
		8=>"EY PA3 TT2",
		9=>"NN1 AA AY NN1",
		10=>"TT2 EH EH NN1",
		11=>"IH LL EH EH VV IH NN1",
		12=>"TT2 WH EH EH LL VV",
		13=>"TH ER1 PA2 PA3 TT2 IY NN1",
		14=>"FF OR PA2 PA3 TT2 IY NN1",
		15=>"FF IH FF PA2 PA3 TT2 IY NN1",
		16=>"SS SS IH PA2 PA3 TT2 IY NN1",
		17=>"SS SS EH VV IH NN1 PA2 PA3 TT2 IY NN1",
		18=>"EY PA2 PA3 TT2 IY NN1",
		19=>"NN1 AY NN1 PA2 PA3 TT2 IY NN1",
		20=>"TT2 WH EH EH NN1 PA2 PA3 TT2 IY",
		30=>"TH ER2 PA2 PA3 tt2 iy",
		40=>"FF OR PA3 TT2 IY",
		50=>" FF IH FF FF PA2 PA3 TT2 IY",
		60=>"SS SS IH PA3 KK2 SS PA2 PA3 TT2 IY",
		70=>"SS SS EH VV IN NN1 PA2 PA3 TT2 IY",
		80=>"EY PA3 TT2 IY",
		90=>"NN1 AY NN1 PA3 TT2 IY",
		100=>"HH2 AX AX NN1 PA2 DD2 RR2 IH IH PA1 DD1",
		1000=>"TH AA AW ZZ TH PA1 PA1 NN1 DD1",
		1000000=>"MM IH IH LL YY1 AX NN1",
		1000000000=>"BB2 IH IH LL YY1 AX NN1",
		"sunday"=>"SS SS AX AX NN1 PA2 DD2 EY",
		"monday"=>"MM AX AX NN1 PA2 DD2 EY",
		"tuesday"=>"TT2 UW ZZ PA2 DD2 EY",
		"wednesday"=>"WW EH EH NN1 ZZ PA2 DD2 EY",
		"thursday"=>"TH ER2 ZZ PA2 DD2 EY",
		"friday"=>"FF RR2 AY PA2 DD2 EY",
		"saturday"=>"SS SS AE PA3 TT2 PA2 DD2 EY",
		"january"=>"JH AE AE NN1 YY2 XR IY",
		"february"=>"FF EH EH PA1 BR RR2 UW2 XR IY",
		"march"=>"MM AR PA3 CH",
		"april"=>"EY PA3 PP RR2 IH IH LL",
		"may"=>"MM EY",
		"june"=>"JH UW2 NN1",
		"july"=>"JH UW1 LL AY",
		"august"=>"AO AO PA2 GG2 AX SS PA3 TT1",
		"september"=>"SS SS EH PA3 PP PA3 TT2 EH EH PA1 BB2 ER1",
		"october"=>"AA PA2 KK2 PA3 TT2 OW PA1 BB2 ER1",
		"november"=>"NN2 OW VV EH EH MM PA1 BB2 ER1",
		"december"=>"DD2 IY SS SS EH EH MM PA1 BB2 ER1",
	};

}

1;


SP0256/lib/Speech/SP0256.pm
package main;
use strict; use warnings;
unless (caller){
	
	my $chip=Speech::SP0256->new();
	my $a=scalar keys %{$chip->{dic}};
	my $r=@{$chip->{rules}};
	print "\n\n  ******  Speech Using Pure Perl Emulating a Retro SP0256   *****\n",
	      "  *                  Version 0.01                               *\n",
	      "  *       Allophones available:                                 *\n",
	      "  *  @{$chip->{items}}[0..15]   *\n",
	      "  *  @{$chip->{items}}[16..31]    *\n",
	      "  *  @{$chip->{items}}[32..47]      *\n",
	      "  *  @{$chip->{items}}[47..63] *\n",   
	      "  *       Words available $a   Rules Available  $r              *\n",
	      "  ***************************************************************\n\n",
	      "   Enter words to speak;\n   Allophones may be passed as /XX/ e.g. /PP/ /ER1/ /LL/:\n\n" ;

	$chip->speak("hello");
	print "\n";
	my $input="";
	while ($input ne "q"){
		$input=<STDIN>;
		chomp $input;
		$chip->speak($input);
		print "\n";
	}
};

package Speech::SP0256;
use strict; use warnings;
use lib "../../lib";
our $VERSION=0.03;
use Speech::SP0256::Allophone;
use Speech::SP0256::Dictionary;
use Speech::SP0256::Rules;
use Storable;
use if $^O eq 'MSWin32', "Win32::Sound";

our $dsp;

sub start{
	my $self=shift;
	if ($^O eq 'MSWin32'){
	   our $dsp = new Win32::Sound::WaveOut(8000, 8, 1);
	} 
	else{
	   open(our $dsp,"|padsp tee /dev/dsp > /dev/null") or warn "DSP can not be intiated $!";
    }	   
}

sub new{
	my $class=shift;	
	our $dsp;
	my $self=Speech::SP0256::Allophone->build();
	Speech::SP0256::Dictionary::initDic($self);
	Speech::SP0256::Rules::setRules($self);
	bless $self, $class;
	return $self;

}

sub debug{
	my ($self, $debug)=@_;
	$self->{debug}=$debug?1:0;
}

sub utter{
	my ($self,$ap)=@_;
	my @sounds=split(/[^a-zA-Z0-9]+/,uc $ap);
	unless ($dsp){
		$self->start();; 
	}
	foreach my $sound(@sounds){
		print $sound," " if $self->{debug};
		next unless $self->{allophones}->{$sound};  # ignore sounds that are not available
		my $b=$self->{allophones}->{$sound}->{s};
		if ($^O eq 'MSWin32'){
		   $dsp->Load($b);       # get it
		   $dsp->Write();           # hear it
	    } 
		else{
	       while (length $b){$b=substr $b,syswrite $dsp,$b};
        }
		
	}
#	print "\n";
}

sub speak{
	my ($self,$text)=@_;
	my $build="";my $nextChar="";
	while (length $text){
		$nextChar=substr $text, 0, 1, "";
		if ($nextChar=~/[ \.,;\-:?]/){
            $self->utter($self->tts($build) ." ".$self->tts($nextChar));
            $build="";
		}
		else{
			$build.=$nextChar;
		}
	}
	$self->utter($self->tts($build));
}

sub tts{
	my ($self,$word)=@_;
	if ($word=~/\/([A-Z]{2}\d?)\//){return $1};    # Direct Allophone output
	my $result=lc $word;             
	if (defined $self->{dic}->{$result}){          # Word is in dictionary
		$result= $self->{dic}->{$result} 
	}
	elsif($word=~/^\d+$/){
		$result=$self->numberSpeak($word); 
	}
	else{
		foreach my $rule (@{$self->{rules}}){      # Build word from set of rules
			my ($match,$replace)=@$rule;
			if ($result=~/$match/){
				$result=~s/$match/ $replace /g;	
				#print "Match $match found, replaced with $replace \n";
			}	
		}
		$result=~s/\s+/ /g;
	}
	print "Word found $word ; output is $result\n" if $self->{debug};
	return $result;
}

sub numberSpeak{
	my ($self,$number)=@_;
	my $split=numberSplit(1*$number);
	print $split,"\n" if $self->{debug};
	my $inWords="";
	foreach (split " ",$split){
		$inWords.=$self->tts($_);
		$inWords.=" PA4 "
	}
	return $inWords;
	
}

sub numberSplit{
	my $number=shift;
	return "0" if ($number== 0);
    if ($number < 0){ return "minus " . numberSplit(-$number)};
    if ($number >= 1000000000){ return numberSplit(int($number / 1000000000)) . " 1000000000 " . numberSplit($number % 1000000000)};
    if ($number >= 1000000){ return numberSplit(int($number / 1000000)) . " 1000000 " . numberSplit($number % 1000000)};
    if ($number >= 1000){ return numberSplit(int($number / 1000)) . " 1000 " . numberSplit($number % 1000)};
    if ($number >= 100){ return numberSplit(int($number / 100)) . " 100 " . numberSplit($number % 100)};
    if ($number >= 20){ return (10*int($number/10))." ".($number%10) };
    return $number if $number;
    return "";
}




1;
